activejob/lib/active_job/concurrency.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+module ActiveJob
+  module Concurrency
+    extend ActiveSupport::Concern
+    extend ActiveSupport::Autoload
+
+    autoload :Limit
+
+    included do
+      class_attribute :_concurrency_limit, instance_accessor: false
+      class_attribute :_concurrency_keys, instance_accessor: false
+      class_attribute :_concurrency_timeout, instance_accessor: false
+    end
+
+    module ClassMethods
+      def concurrency_limit
+        self._concurrency_limit
+      end
+
+      def concurrency_keys
+        self._concurrency_keys
+      end
+
+      def concurrency_timeout
+        self._concurrency_timeout
+      end
+
+      def concurrency(limit:, keys: [], timeout: 120)
+        self._concurrency_limit = limit
+        self._concurrency_keys = keys
+        self._concurrency_timeout = timeout
+      end
+    end
+
+    attr_accessor :concurrency_limit
+
+    attr_accessor :concurrency_keys
+
+    attr_accessor :concurrency_timeout
+
+    def serialize
+      super.merge(
+        "concurrency_limit"   => concurrency_limit,
+        "concurrency_keys"    => concurrency_keys,
+        "concurrency_timeout" => concurrency_timeout
+      )
+    end
+
+    def deserialize(job_data)
+      super
+      self.concurrency_limit   = job_data["concurrency_limit"]
+      self.concurrency_keys    = job_data["concurrency_keys"]
+      self.concurrency_timeout = job_data["concurrency_timeout"]
+    end
+
+    def concurrency_limit
+      self.class.concurrency_limit
+    end
+
+    def concurrency_keys
+      self.class.concurrency_keys
+    end
+
+    def concurrency_timeout
+      self.class.concurrency_timeout
+    end
+
+    def concurrency_limit_instance
+      @concurrency_limit_instance ||= Limit.new(self.class.concurrency_limit)
+    end
```

Comment:
Only used for the adapters before enqueuing the job. The instance is not meant to be serialized with the job.

Reply:
Do we need to cache the instance?  Since jobs can be serialized / deserialized, I'd rather we don't keep references to things unless absolutely necessary.  This allocation is probably not a bottleneck, so I'd prefer we don't cache unless there is a reason.